### ISSUE TITLE
MSBuild TaskLoggingHelper is not respecting \r

### ISSUE THREAD

**Issue by mishra14** [Microsoft/msbuild#2316]
## Repro steps - 

dotnet new console
dotnet add package newtonsoft.json
edit csproj to change the newtonsoft versionto 12.0.0
run msbuild /t:Restore on the project

## Expected -
Shows errors

## Actual -
Shows errors with \r at the end of the line 

```
"C:\Users\anmishr\Source\Repos\ConsoleApp21\ConsoleApp21\ConsoleApp21.csproj" (restore target) (1) ->
(Restore target) ->
  E:\NuGet.Client\src\NuGet.Core\NuGet.Build.Tasks\NuGet.targets(102,5): error : TESTING restore\r [C:\Users\anmishr\Source\Repos\ConsoleApp21\ConsoleApp21\ConsoleApp21.csproj]
E:\NuGet.Client\src\NuGet.Core\NuGet.Build.Tasks\NuGet.targets(102,5): error : this is new line [C:\Users\anmishr\Source\Repos\ConsoleApp21\ConsoleApp21\ConsoleApp21.csproj]
  C:\Users\anmishr\Source\Repos\ConsoleApp21\ConsoleApp21\ConsoleApp21.csproj : error NU1102: Unable to find package newtonsoft.json with version (>= 12.0.0)\r
C:\Users\anmishr\Source\Repos\ConsoleApp21\ConsoleApp21\ConsoleApp21.csproj : error NU1102:   - Found 57 version(s) in nuget.org [ Nearest version: 10.0.3 ]\r
```

## Notes - 

Here NuGet is sending messages through [TaskHelper.LogError](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks/Common/MSBuildLogger.cs#L87). I have verified that the messages contain both \r\n as is. So it doesn't seem to be an issue with NuGet. 

## Versions - 
dotnet cli - 2.0.0-preview2
msbuild - Microsoft (R) Build Engine version 15.3.406.54721 for .NET Framework

**Comment by mishra14** [comment]
cc: @rainersigwald @emgarten

**Comment by rainersigwald** [comment]
This happens because the log event (with an embedded CRLF) filters down to a point where CR is [explicitly filtered out](https://github.com/Microsoft/msbuild/blob/07c3b3392dc05e0e3aba18bdec235a374aa72301/src/Shared/EventArgsFormatting.cs#L74) if the current log output device [is a character device](https://github.com/Microsoft/msbuild/blob/07c3b3392dc05e0e3aba18bdec235a374aa72301/src/Build/Logging/BaseConsoleLogger.cs#L302-L323) like the console. The filtering works [by escaping `\r` to `\\r`](https://github.com/Microsoft/msbuild/blob/07c3b3392dc05e0e3aba18bdec235a374aa72301/src/Shared/EventArgsFormatting.cs#L23-L31).

That's a pain for a multi-line error message that wants to just use `Environment.NewLine`, because within the task that's emitting an error, you can't know what loggers are attached. You'd like to emit `\r\n` to be friendly to text logs, but it produces this artifact for console logs.

**Comment by rainersigwald** [comment]
It's worse than I thought. There's code that [splits the user message on newlines](https://github.com/Microsoft/msbuild/blob/8c48a368a1aec5cba34b473265a712e28b948701/src/Shared/EventArgsFormatting.cs#L330), [reassembles it back using our preferred newlines](https://github.com/Microsoft/msbuild/blob/8c48a368a1aec5cba34b473265a712e28b948701/src/Shared/EventArgsFormatting.cs#L333-L346), returns as a string, then [splits again](https://github.com/Microsoft/msbuild/blob/8c48a368a1aec5cba34b473265a712e28b948701/src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs#L1271) (using a [_different_ split method](https://github.com/Microsoft/msbuild/blob/8c48a368a1aec5cba34b473265a712e28b948701/src/Build/Logging/BaseConsoleLogger.cs#L175-L178)), and [emits a log message for each line](https://github.com/Microsoft/msbuild/blob/8c48a368a1aec5cba34b473265a712e28b948701/src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs#L1303), possibly [split across multiple lines](https://github.com/Microsoft/msbuild/blob/8c48a368a1aec5cba34b473265a712e28b948701/src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs#L1284-L1297).

So there's a bunch of unnecessary allocations happening and ALSO it would be totally fine to call `LogError(stringWithEmbeddedCRLF)`.

I'm removing all of the escaping stuff, but not currently tackling the split-join-split problem.

**Comment by mishra14** [comment]
@rainersigwald thanks for digging deeper in this. Do we have a timeline for fixing this?

**Comment by rainersigwald** [comment]
@mishra14 I've got #2325 out to get this fixed but it's in our 15.5 branch. Based on mail, we don't think it's a good idea to take the change right now--I'm pretty sure I understand all the implications but we're pretty late in the cycle.
